cloudevents/pkg/reconciler/events/cloudevent/cloudevent.go

```diff
@@ -0,0 +1,162 @@
+package cloudevent
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	cdeevents "github.com/cdfoundation/sig-events/cde/sdk/go/pkg/cdf/events"
+	cloudevents "github.com/cloudevents/sdk-go/v2"
+	"github.com/google/uuid"
+	"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1"
+	"github.com/tektoncd/pipeline/pkg/reconciler/events/cloudevent"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"knative.dev/pkg/apis"
+	"knative.dev/pkg/controller"
+	"knative.dev/pkg/logging"
+	"time"
+)
+
+// CEClient matches the `Client` interface from github.com/cloudevents/sdk-go/v2/cloudevents
+type CEClient cloudevents.Client
+
+// TektonCloudEventData type is used to marshal and unmarshal the payload of
+// a Tekton cloud event. It can include a TaskRun or a PipelineRun
+type TektonCloudEventData struct {
+	TaskRun     *v1beta1.TaskRun     `json:"taskRun,omitempty"`
+	PipelineRun *v1beta1.PipelineRun `json:"pipelineRun,omitempty"`
+}
+
+// newTektonCloudEventData returns a new instance of TektonCloudEventData
+func newTektonCloudEventData(runObject objectWithCondition) TektonCloudEventData {
+	tektonCloudEventData := TektonCloudEventData{}
+	switch v := runObject.(type) {
+	case *v1beta1.TaskRun:
+		tektonCloudEventData.TaskRun = v
+	case *v1beta1.PipelineRun:
+		tektonCloudEventData.PipelineRun = v
+	}
+	return tektonCloudEventData
+}
+
+func getEventType(runObject objectWithCondition) (*cdeevents.CDEventType, error) {
+	c := runObject.GetStatusCondition().GetCondition(apis.ConditionSucceeded)
+	if c == nil {
+		return nil, fmt.Errorf("no condition for ConditionSucceeded in %T", runObject)
+	}
+	var eventType cdeevents.CDEventType
+	switch {
+	// TODO : Update TaskRun/PipelineRun events from finished to success/fa
+	case c.IsUnknown():
+		switch runObject.(type) {
+		case *v1beta1.TaskRun:
+			switch c.Reason {
+			case v1beta1.TaskRunReasonStarted.String():
+				eventType = cdeevents.TaskRunStartedEventV1
+				//case v1beta1.TaskRunReasonRunning.String():
+				//	eventType = TaskRunRunningEventV1
+				//default:
+				//	eventType = TaskRunUnknownEventV1
+			}
+		case *v1beta1.PipelineRun:
+			switch c.Reason {
+			case v1beta1.PipelineRunReasonStarted.String():
+				eventType = cdeevents.PipelineRunStartedEventV1
+				//case v1beta1.PipelineRunReasonRunning.String():
+				//	eventType = PipelineRunRunningEventV1
+				//default:
+				//	eventType = PipelineRunUnknownEventV1
+			}
```

Comment:
Since the controller is notified here when the `PipelineRun` is created (as opposed to when it's picked up by the other controller), we could in fact send the "Queued" event here, and send the "Started" event for the Tekton "Running" condition, which corresponds to the first TaskRun being created. wdyt?

Reply:
That association makes a lot of sense
Started -> CDE Queued
Running -> CDE Started
considering there is no "Running" in the CDE events Started means that it has started Running.